extensions/documents/homie meta extension.md

```diff
@@ -0,0 +1,121 @@
+# Meta
+
+Version: **<!--VERSION-->1.0.0<!--VERSION-->**
+Date: **<!--DATE-->10. Jul 2019<!--DATE-->**
+Authors: **<!--AUTHORS-->[EPNW](https://epnw.eu)<!--AUTHORS-->**
+License: **<!--LICENSE-->[CCA 4.0](https://homieiot.github.io/license)<!--LICENSE-->**
+
+## Abstract
+This extension defines how to add metadata and tags to devices, nodes and properties.
+
+Tags are simple annotations that ever device, node oder property can have.
+Metadata on the other hand are more complex. They allow the definition of multiple *mainkeys* and *mainvalues* for each device, node or property.
+Each main-key-value-pair may have nested sub-key-value-pairs.
+Having metadata might be useful for Homie controllers.
+A concrete usecase is the openHAB controller, which relies on metadata and tags to make voice assistants like Amazons Alexa or the Google Assistant able to control it.
+With this extension, properties are able to advertise how they want to be treated.
+
+## Extension Identifier
+The ID of this extension is `eu.epnw.meta`.
+
+## Homie Version
+This extension supports Homie `3.0.1` and `4.0`.
+
+## Extension Datatypes
+This extension defines no new datatypes.
+
+## Extension Attributes
+
+### Extension Attributes for any kind of Items
+
+The following attributes are all **optional** and may be added to any kind of item (devices, nodes or properties).
+
+| Topic                | Description                            | Payload type                         | Default value|
+|----------------------|----------------------------------------|--------------------------------------|--------------|
+| $tags                | List of *tags* for the item            | Comma (`,`) separated list of *tags* |              |
+| $meta/$mainkey-count | The number of *mainkeys* this item has | Positive Integer                     | 0            |
+
+
+Since *tags* are represented by a comma separated list, a limitation is that a tag itselfe MUST NOT contain a comma.
+
+**Examples**
+Assuming the base topic is *homie*, device ID is *super-car*, the node is *engine* and the property is *temperature* then:
+```java
+homie/super-car/engine/temperature/$meta/$key-count → "2"  
```

Comment:
You are right, I haven't thought about that! My first approach was to use a list, and then have the subtopics accessed by their keynames instead of indexes. E.g:
```java
homie/super-car/engine/temperature/$meta/$keys → "alexa,homekit" 
homie/super-car/engine/temperature/$meta/alexa/$value → "Fan" 
```
But then I noticed that this is a very strong limitation, since the keys are then used as topic IDs. And topic IDs "MAY contain lowercase letters from a to z, numbers from 0 to 9 as well as the hyphen character (-)" (quote from the [covention](https://homieiot.github.io/specification/)).
I guess a solution would be to make the indexing still array like, but have a $keys attribute like in the example above. Then
```java
homie/super-car/engine/temperature/$meta/1/$key → "alexa" 
homie/super-car/engine/temperature/$meta/1/$value → "Fan" 
```
is only valid, if the keys list contains `alexa`.

Reply:
> keys are then used as topic IDs

That's a reasonable limitation. Remember that you can always base48 or base64 encode a key

Reply:
I redesigned the key addressing. I choose a list as you suggested instead of a count.
But to avoid the limitation regarding the topic IDs, the lists elements are not key names, but just key IDs. The key name and value are then nested topics under the keys topic ID.
I did not like the idea of encoding the key name base64 and use it as topic ID, since then the key name itself is not human readable (and base64 contains upper case letters and `=`,  which are not valid topic characters).